python_modules/dagster/dagster_tests/asset_defs_tests/test_asset_selection.py

```diff
@@ -141,6 +141,10 @@ def test_asset_selection_key_prefixes(all_assets: _AssetList):
     sel = AssetSelection.key_prefixes("plants")
     assert sel.resolve(all_assets) == _asset_keys_of(set())
 
+    # works for source assets
+    sel = AssetSelection.key_prefixes("celestial")
```

Comment:
If I understand correctly including source assets with `key_prefixes` is inconsistent with the behavior of `groups`? I would expect them to have the same behavior with respect to source assets.

Reply:
I think it's a wash between similarity to `groups` (no source assets) or `keys` (yes source assets). Erring on the side of `keys` since it's better to have more ways to select source assets (the reason `groups` doesn't is just for backcompat). That said, I can change if you feel strongly.

Reply:
>Isn't it also because we don't want to allow people to select both source and non-source assets at once? Or does that fall under backcompat?

Well, at the time `AssetSelection` was designed, there was no user code associated with a source asset, no it didn't make sense for `AssetSelection` to capture them-- that's why none of the older `AssetSelection` methods select source assets.

But in a (future) world where source asset observations can run together with materializing assets, then it makes sense to be able to select them together. IMO the only reason we wouldn't do so for `groups` immediately when this functionality is available is that it would change existing jobs defined with this method-- so, yes, purely backcompat.

Reply:
> But in a (future) world where source asset observations can run together with materializing assets, then it makes sense to be able to select them together.

Got it, this makes sense.

In this future, do you have thoughts on how we'd want users to be able to specify "only the observable assets within this group" or "only the materializable assets within this group" or vice versa?

> IMO the only reason we wouldn't do so for groups immediately when this functionality is available is that it would change existing jobs defined with this method-- so, yes, purely backcompat.

FWIW, I think we're technically allowed to do this because observable source assets are experimental? Or are there situations where this would affect how non-observable source assets are handled?  E.g. it could affect the behavior of the `monitored_assets` argument of `multi_asset_sensor`. Although `multi_asset_sensor` is experimental too.

Reply:
>In this future, do you have thoughts on how we'd want users to be able to specify "only the observable assets within this group" or "only the materializable assets within this group" or vice versa?

I've added `include_sources` kwargs that default to `False` for both `key_prefixes` and `groups`-- that makes them behave similarly as filters and preserves backcompat.

>FWIW, I think we're technically allowed to do this because observable source assets are experimental? Or are there situations where this would affect how non-observable source assets are handled?

This will cause job definitions that currently don't execute observable source assets and include nothing experimental to start executing them, so it doesn't seem like it falls under the "we can do it because it's experimental" rule.